samples/mpu/mpu_stack_guard_test/src/main.c

```diff
@@ -10,62 +10,79 @@
 #include <misc/printk.h>
 
 /* size of stack area used by each thread */
-#define STACKSIZE 1024
+#define STACKSIZE  512
 
 /* scheduling priority used by each thread */
 #define PRIORITY 7
 
 /* delay between greetings (in ms) */
 #define SLEEPTIME 500
 
+/* delay between greetings (in ms) */
+#define SCHEDTIME 30
+
 /* Focaccia tastes better than DEEDBEEF ;-) */
 #define STACK_GUARD_CANARY 0xF0CACC1A
 
 struct stack_guard_buffer {
 	/* Make sure canary is not optimized by the compiler */
+	struct k_thread thread;
 	volatile u32_t canary;
 	K_THREAD_STACK_MEMBER(stack, STACKSIZE);
 };
 
-struct stack_guard_buffer buf;
-struct k_thread test_thread;
+#define LAST_THREAD_NUM 3
+struct stack_guard_buffer buf[LAST_THREAD_NUM+1];
 
-u32_t recursive_loop(u32_t counter)
+u32_t recursive_loop(u32_t counter, int num, void *dummy)
 {
-	if (buf.canary != STACK_GUARD_CANARY) {
-		printk("Canary = 0x%08x\tTest not passed.\n", buf.canary);
+
+	if (buf[num].canary != STACK_GUARD_CANARY) {
+		printk("Canary = 0x%08x\tTest not passed.\n", buf[num].canary);
 
 		while (1) {
 			k_sleep(SLEEPTIME);
 		}
 	}
-
+	k_sleep(SCHEDTIME);
 	counter++;
-
-	return recursive_loop(counter) + recursive_loop(counter);
+	if (dummy == 0)
+		return counter;
+	return recursive_loop(counter, num, dummy)
+		+recursive_loop(counter, num, dummy);
 }
 
+
 /* stack_guard_thread is a dynamic thread */
 void stack_guard_thread(void *dummy1, void *dummy2, void *dummy3)
 {
 	ARG_UNUSED(dummy1);
 	ARG_UNUSED(dummy2);
 	ARG_UNUSED(dummy3);
+	while (1)
+		recursive_loop(0, (int)dummy1, dummy2);
 
-	u32_t result = recursive_loop(0);
-
-	/* We will never arrive here */
-	printk("Result: %d", result);
 }
 
 void main(void)
 {
-	buf.canary = STACK_GUARD_CANARY;
+	int i;
+	bool failed;
+
+	printk("STACK_ALIGN %x\n", STACK_ALIGN);
+	for (i = 0; i <= LAST_THREAD_NUM; i++) {
+		buf[i].canary = STACK_GUARD_CANARY;
+		failed = (i == LAST_THREAD_NUM) ? true : false;
+		if (failed) {
+			printk("MPU STACK GUARD Test\n");
+			printk("Canary Initial Value = 0x%x threads %p\n",
+					buf[i].canary, &buf[i].thread);
+		}
 
-	printk("MPU STACK GUARD Test\n");
-	printk("Canary Initial Value = 0x%x\n", buf.canary);
+		/* create stack_guard_thread */
+		k_thread_create(&buf[i].thread, buf[i].stack, STACKSIZE,
```

Comment:
yeah.  That adjusts for the alignment adjustment.  Otherwise you might get unlucky and bad things will happen.  Good catch.